parlai/agents/image_seq2seq/image_seq2seq.py

```diff
@@ -0,0 +1,349 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Image+Seq2Seq Agent.
+"""
+import torch
+from typing import Dict, List, Tuple
+
+from .modules import ImageSeq2seqModel
+from parlai.agents.transformer.transformer import TransformerGeneratorAgent
+from parlai.core.dict import DictionaryAgent
+from parlai.core.message import Message
+from parlai.core.torch_agent import Batch, Output
+from parlai.utils.misc import round_sigfigs, Opt
+
+try:
+    from nltk.translate import bleu_score as nltkbleu
+
+except ImportError:
+    nltkbleu = None
+
+try:
+    from fairseq import bleu as fairseq_bleu
+
+except ImportError:
+    fairseq_bleu = None
+
+TOKEN_IMAGE = '__image__'
+TOKEN_NO_IMAGE = '__no_image__'
+
+
+class ImageSeq2seqAgent(TransformerGeneratorAgent):
+    """
+    ImageSeq2seqAgent Agent.
+
+    Combines a transformer generator with images.
+    """
+
+    def __init__(self, opt: Opt, shared: dict = None):
+        super().__init__(opt, shared)
+        self.compute_tokenized_bleu = opt.get('compute_tokenized_bleu', False)
+        if not shared:
+            self._init_bleu_scorers()
+        else:
+            self.fairseq_bleu_scorer = shared['fairseq_bleu_scorer']
+            self.nltk_bleu = shared['nltk_bleu']
+            self.nltk_bleu_cnts = shared['nltk_bleu_cnts']
+
+    def _init_bleu_scorers(self):
+        if not hasattr(self, 'fairseq_bleu_scorer'):
+            if fairseq_bleu is None:
+                self.fairseq_bleu_scorer = None
+            else:
+                self.fairseq_bleu_scorer = fairseq_bleu.Scorer(
+                    self.NULL_IDX, self.END_IDX, self.dict[self.dict.unk_token]
+                )
+        self.nltk_bleu = {f'bleu-{i}': 0 for i in range(1, 5)}
+        self.nltk_bleu_cnts = {f'bleu-{i}': 0 for i in range(1, 5)}
+
+    def share(self) -> dict:
+        """
+        Override to include BLEU scorer.
+        """
+        shared = super().share()
+        shared['fairseq_bleu_scorer'] = self.fairseq_bleu_scorer
+        shared['nltk_bleu'] = self.nltk_bleu
+        shared['nltk_bleu_cnts'] = self.nltk_bleu_cnts
+        return shared
+
+    def build_model(self) -> ImageSeq2seqModel:
+        """
+        Override to build appropriate model.
+        """
+        self.model = ImageSeq2seqModel(self.opt, self.dict)
+        if self.opt['embedding_type'] != 'random':
```

Comment:
this should be taken care of by `copy_embeddings` in torch_agent

Reply:
i think i actually do need this here, as it's taken care of in `TransformerGeneratorAgent.build_model` (which I subclass); I wasn't able to find `copy_embeddings` in regular TA

Reply:
So @EricMichaelSmith's tutorial on TGA actually covers the recommended usage:

https://github.com/facebookresearch/ParlAI/blob/45e7516ef4ed54d8a70a39cf32d10b529a8250bb/parlai/agents/examples/seq2seq.py#L192-L201